src/client/csfle/state_machine.rs

```diff
@@ -0,0 +1,171 @@
+use std::convert::TryInto;
+
+use bson::{Document, RawDocument, RawDocumentBuf};
+use futures_util::{stream, TryStreamExt};
+use mongocrypt::ctx::{Ctx, State};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    sync::oneshot,
+};
+
+use crate::{
+    client::options::ServerAddress,
+    cmap::options::StreamOptions,
+    error::{Error, Result},
+    operation::{RawOutput, RunCommand},
+    runtime::AsyncStream,
+    Client,
+};
+
+impl Client {
+    pub(crate) async fn run_mongocrypt_ctx(
+        &self,
+        ctx: Ctx,
+        db: Option<&str>,
+    ) -> Result<RawDocumentBuf> {
+        let guard = self.inner.csfle.read().await;
+        let csfle = match guard.as_ref() {
+            Some(csfle) => csfle,
+            None => return Err(Error::internal("no csfle state for mongocrypt ctx")),
+        };
+        let mut result = None;
+        // This needs to be a `Result` so that the `Ctx` can be temporarily owned by the processing
+        // thread for crypto finalization.  An `Option` would also work here, but `Result` means we
+        // can return a helpful error if things get into a broken state rather than panicing.
+        let mut ctx = Ok(ctx);
+        loop {
+            let state = result_ref(&ctx)?.state()?;
+            match state {
+                State::NeedMongoCollinfo => {
+                    let ctx = result_mut(&mut ctx)?;
+                    let filter = raw_to_doc(ctx.mongo_op()?)?;
+                    let metadata_client = csfle
+                        .aux_clients
+                        .metadata_client
+                        .as_ref()
+                        .and_then(|w| w.upgrade())
```

Comment:
I know this was a prior PR, but where do the strong references for these live again? Is the user in charge of ensuring they're alive?